### PR TITLE
Add support for signed uploads.

### PR DESCRIPTION
**Pull request by dmitry-mukhin**
## Description

Add support for signed uploads, using new HMAC-SHA256 method.

## Checklist

- [x] Tests (if applicable)
- [ ] Documentation (if applicable)
- [x] Changelog stub (or use [conventional commit messages](https://www.conventionalcommits.org/))

### CODE REVIEW

File: pyuploadcare/api.py [hunk 1]
```diff
@@ -66,7 +66,8 @@ def _build_user_agent():
 
 
 def _signed_upload_signature(secret, expire):
-    return hmac.new(str(secret), str(expire), hashlib.sha256).hexdigest()
+    k, m = map(lambda x: str(x).encode(), (secret, expire))
```
Comment (andrewshkovskii) [inline pyuploadcare/api.py:69]:
```suggestion
    k, m = str(secret).encode(), str(expire).encode()
```

File: tests/functional/test_api.py [hunk 2]
```diff
@@ -68,3 +69,28 @@ def test_options(self, request):
...
+
+        request.return_value = MockResponse(200, b'[]')
+
+        uploading_request('GET', '')
+        data = request.call_args[1]['data']
+        self.assertIn('expire', data)
+        self.assertIn('signature', data)
+
+    @patch('requests.sessions.Session.request', autospec=True)
+    def test_signature_fields_scipped(self, request):
```
Comment (andrewshkovskii) [inline tests/functional/test_api.py:90]: Typo? `scipped`

File: tests/functional/test_api.py [hunk 3]
```diff
@@ -68,3 +69,28 @@ def test_options(self, request):
...
+        conf.signed_uploads = True
+
+        request.return_value = MockResponse(200, b'[]')
+
+        uploading_request('GET', '')
+        data = request.call_args[1]['data']
+        self.assertIn('expire', data)
+        self.assertIn('signature', data)
+
+    @patch('requests.sessions.Session.request', autospec=True)
```
Comment (andrewshkovskii) [inline tests/functional/test_api.py:89]:
```suggestion
```

File: tests/functional/test_api.py [hunk 4]
```diff
@@ -68,3 +69,28 @@ def test_options(self, request):
         request.return_value = MockResponse(200, b'')
 
         rest_request('OPTIONS', 'files/')
+
+
+class SignatureTest(unittest.TestCase):
+    def tearDown(self):
+        conf.signed_uploads = False
+
+    @patch('requests.sessions.Session.request', autospec=True)
```
Comment (andrewshkovskii) [inline tests/functional/test_api.py:78]:
```suggestion
```

File: tests/functional/test_api.py [hunk 5]
```diff
@@ -68,3 +69,28 @@ def test_options(self, request):
         request.return_value = MockResponse(200, b'')
 
         rest_request('OPTIONS', 'files/')
+
+
```
Comment (andrewshkovskii) [inline tests/functional/test_api.py:73]:
```suggestion
@patch('requests.sessions.Session.request', autospec=True)
```

File: tests/functional/test_api.py (hunk 2 above)
Comment (andrewshkovskii) [inline tests/functional/test_api.py:90]: What value of `conf.signed_uploads` should be in this test?

Comment (dmitry-mukhin) [reply tests/functional/test_api.py:89]: default is False.

File: tests/integration/test_api_resources.py [hunk 6]
```diff
@@ -107,6 +109,35 @@ def test_successful_upload_from_url_sync_with_filename(self):
...
+class SignedUploadsTest(RestAPITestCase):
+    def setUp(self):
+        conf.pub_key = os.environ.get("SIGNED_UPLOADS_PUBKEY")
+        conf.secret = os.environ.get("SIGNED_UPLOADS_SECKEY")
+        conf.signed_uploads = True
+
+    def tearDown(self):
+        conf.pub_key = None
+        conf.secret = None
+        conf.signed_uploads = False
```
Comment (andrewshkovskii) [inline tests/integration/test_api_resources.py:123]: I think we could use mock's for this. `mock.patch.object` as class decorators and there will be no need to add `tearDown` and `setUp` methods.

Comment (dmitry-mukhin) [reply tests/integration/test_api_resources.py:123]: may be later :)

File: pyuploadcare/api.py [hunk 7]
```diff
@@ -65,6 +65,11 @@ def _build_user_agent():
                                                       extension_info)
 
 
+def _signed_upload_signature(secret, expire):
+    k, m = str(secret).encode(), str(expire).encode()
```
Comment (homm) [inline pyuploadcare/api.py:69]: What does `str(secret)` mean? This is not a proper way to convert bytes to string:

```python
>>> str(b'0123456789abcdef')
<<< "b'0123456789abcdef'"
```

**Comment by andrewshkovskii** [comment]
And then I wonder how tests are passed for py >3? - there are no integration tests for Travis builds. Why, @dmitry-mukhin ?
Alex is right about invalid bytes-to-string conversion

File: pyuploadcare/api.py (hunk 7 above)
Comment (dmitry-mukhin) [reply pyuploadcare/api.py:69]: we don't need bytes to string, we need vice versa for hmac.
this was caught by tests on py3 @andrewshkovskii

**Comment by dmitry-mukhin** [comment]
@andrewshkovskii [there are](https://github.com/uploadcare/pyuploadcare/blob/master/.travis.yml#L105-L111)

File: pyuploadcare/api.py (hunk 7 above)
Comment (homm) [reply pyuploadcare/api.py:69]: > we don't need bytes to string

What does `str(secret)` mean?

Comment (dmitry-mukhin) [reply pyuploadcare/api.py:69]: we want to cast user's input to `string` (e.g. `None` or numbers) and `expire` is always a number

Comment (homm) [reply pyuploadcare/api.py:69]: Ok, this makes sence for `expire`, but leads to difficult to find errors if the `secret` is already bytes.

**Pull request merged**